src/test/scala/com/fulcrumgenomics/umi/DuplexConsensusCallerTest.scala

```diff
@@ -0,0 +1,265 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.testing.SamRecordSetBuilder.{Minus, Plus}
+import com.fulcrumgenomics.testing.{SamRecordSetBuilder, UnitSpec}
+import com.fulcrumgenomics.util.NumericTypes.PhredScore
+
+class DuplexConsensusCallerTest extends UnitSpec {
+  // Function to create a caller without so much writing
+  def caller(q: Int = 10, pre: Int = DuplexConsensusCaller.ErrorRatePreUmi, post: Int = DuplexConsensusCaller.ErrorRatePostUmi) =
+    new DuplexConsensusCaller(readNamePrefix="test", minInputBaseQuality=q.toByte, errorRatePreUmi=pre.toByte, errorRatePostUmi=post.toByte)
+
+  // A default caller that can be used by tests that don't care about the parameters
+  private val c = caller()
+
+  private val MI = ConsensusTags.MolecularId
+
+  "DuplexConsensusCaller.sourceMoleculeId" should "strip of that last /suffix" in {
+    val builder = new SamRecordSetBuilder()
+    builder.addFrag(start=1, attrs=Map(MI -> "foo/A")).foreach { r => c.sourceMoleculeId(r) shouldBe "foo" }
+    builder.addFrag(start=1, attrs=Map(MI -> "f/A/B")).foreach { r => c.sourceMoleculeId(r) shouldBe "f/A" }
+    builder.addFrag(start=1, attrs=Map(MI -> "f/abc")).foreach { r => c.sourceMoleculeId(r) shouldBe "f"   }
+    builder.addFrag(start=1, attrs=Map(MI -> "f///A")).foreach { r => c.sourceMoleculeId(r) shouldBe "f//" }
+  }
+
+  it should "throw an exception if the MI tag doesn't have a /suffix" in {
+    an[Exception] shouldBe thrownBy {
+      val builder = new SamRecordSetBuilder()
+      builder.addFrag(start=1, attrs=Map(MI -> "foo")).foreach { r => c.sourceMoleculeId(r) }
+    }
+  }
+
+  it should "throw an exception if the MI tag is null" in {
+    an[Exception] shouldBe thrownBy {
+      val builder = new SamRecordSetBuilder()
+      builder.addFrag(start=1).foreach { r => c.sourceMoleculeId(r) }
+    }
+  }
+
+  "DuplexConsensusCaller.consensusReadsFromSamRecords" should "not create records from fragments" in {
+    val builder = new SamRecordSetBuilder(readLength=10)
+    builder.addFrag(start=1, attrs=Map(MI -> "foo/A")).foreach(r => r.setReadString("AAAAAAAAAA"))
+    builder.addFrag(start=1, attrs=Map(MI -> "foo/B")).foreach(r => r.setReadString("AAAAAAAAAA"))
+    val recs = c.consensusReadsFromSamRecords(builder.toSeq)
+    recs shouldBe empty
```

Comment:
I seem to get yelled at by the compiler when I try that, hence fishing around and finding this.